### PR TITLE
restructure data production with process functions

### PR DESCRIPTION
**Pull request by JulienDoerner**
With this PR the functions of the data repository are structured in single process functions. 
This happens in preparation for the documentation of custom photon fields. 

The calculation of the data files is not affected. 

Additionaly I added more comments and docstrings for the existing functions.

### CODE REVIEW

**Comment by JulienDoerner** [comment]
I have also moved the folder fieldDensity which was introduced in #15 to a temp folder. This data are only needed for the calculation of the interaction rate and not needed in the simulation. Therefore it should not be stored in the data folder and downloaded by the user.

File: calc_decay.py [hunk 1]
```diff
@@ -9,6 +9,7 @@
 
 class Decay:
     def load(self, s):
+        """ extract decay parameter form a given line of the data file. """
```
Comment (lukasmerten) [inline calc_decay.py:12]: from

File: calc_elasticscattering.py [hunk 2]
```diff
@@ -6,43 +6,38 @@
...
-xs = np.array([iR.romb_pad_zero(x, 513) for x in data['xs']]) * 1e-31
 
 
-for field in fields:
-    print(field.name)
 
     # calculate the interaction rate, averaged over all isotopes
     rate = np.mean([iR.calc_rate_eps(eps, x, gamma, field) for x in xs], axis=0)
-    fname = folder + '/rate_%s.txt' % field.name.split('_')[0]
+    fname = folder + '/rate_%s.txt' % field.name[:3]
```
Comment (lukasmerten) [inline calc_elasticscattering.py:40]: Why did you change the from splitting at "_" to picking the first 3 characters?

File: calc_photodisintegration.py [hunk 3]
```diff
@@ -121,18 +106,36 @@
...
-    photonField.EBL_Gilmore12(),
-    photonField.URB_Protheroe96()
-    ]
+def processEmission(field):
+    """ calculate photon emission probabilities """
 
-for field in fields:
-    print(field.name)
+    folder = "data/Photodisintegration/"
+    if not os.path.isdir:
```
Comment (lukasmerten) [inline calc_photodisintegration.py:137]: Why can't we use the test as in the oher files?
`if not os.path.exists(folder):`


Maybe I just don't get what you want to check here...

File: photonField.py [hunk 4]
```diff
@@ -185,7 +199,14 @@ def __init__(self):
             self.data[z] = eps[::-1], n[i][::-1]
 
 class EBL_Dominguez11(EBL):
+    """ IRB model from Dominguez 2011 """
+
     def __init__(self, which='best'):
+        """ Constructor
+        
+        Input:
+          which : \"best\" for the best fit model, \"upper\" or \"lower\" for the upper/lower uncertaincy.
```
Comment (lukasmerten) [inline photonField.py:208]: uncertainty

File: photonField.py [hunk 5]
```diff
@@ -212,7 +233,15 @@ def __init__(self, which='best'):
...
 
 class EBL_Stecker16(EBL):
+    """ IRB model from Stecker 2016 """
+
     def __init__(self, which='upper'):
+        """
+        Constructor 
+        
+        Input
+          which : \"upper\" or \"lower\" for an estimation of the uncertaincy
```
Comment (lukasmerten) [inline photonField.py:243]: uncertainty

**Comment by lukasmerten** [comment]
Hi Julien
I will check out the branch locally and run some additional tests. However, this looks overall good to me.
Maybe we can add a small bash script that runs all python scripts with the defaul CRPropa settings.

File: calc_elasticscattering.py (hunk 2 above)
Comment (JulienDoerner) [reply calc_elasticscattering.py:40]: The current implementation in CRPropa takes only 3 characters. I changed this to be consistent in the case that users does not stick to the naming with an abbreviation of 3 characters.

File: calc_photodisintegration.py (hunk 3 above)
Comment (JulienDoerner) [reply calc_photodisintegration.py:137]: In this context this would be the same test. The function `isdir` also checkes wether it is a file or a directory. But in this case it is already defined by the folder naming, that it should be a folder.

**Comment by JulienDoerner** [comment]
Hey @lukasmerten, 
I updated the typos. 

For the execution in one script I would preffere a single python file which can run all others, as this does not depend on the shell environment. This could be interesting for users with Windows and macOS systems.

**Comment by lukasmerten** [comment]
Sounds good. I take a look at this next Monday.

File: calc_decay.py [hunk 6]
```diff
@@ -9,6 +9,7 @@
 
 class Decay:
     def load(self, s):
+        """ extract decay parameter form a given line from the data file. """
```
Comment (rafaelab) [inline calc_decay.py:12]: typo: form => from

File: calc_electromagnetic.py [hunk 7]
```diff
@@ -150,25 +158,25 @@ def process(sigma, field, name):
...
-    process(sigmaTPP, field, 'EMTripletPairProduction')
-    process(sigmaICS, field, 'EMInverseComptonScattering')
+if __name__ == "__main__":
+    fields = [
+        photonField.CMB(),
+        photonField.EBL_Kneiske04(),
+        photonField.EBL_Stecker05(),
+        photonField.EBL_Franceschini08(),
+        photonField.EBL_Finke10(),
+        photonField.EBL_Dominguez11(),
```
Comment (rafaelab) [inline calc_electromagnetic.py:168]: Dominguez also has upper/lower limits. This was lost some time ago. Maybe we can use this PR to bring it back: `photonField.EBL_Dominguez11('lower')` (and upper)?

File: calc_pairproduction.py [hunk 8]
```diff
@@ -115,23 +99,45 @@ def integrand(logk, gamma, field):
...
+    np.savetxt('data/ElectronPairProduction/spectrum_IRB.txt', A3, fmt='%.5e')
+
+
+    fields = [
+        photonField.CMB(),
+        photonField.EBL_Kneiske04(),
+        photonField.EBL_Stecker05(),
+        photonField.EBL_Franceschini08(),
+        photonField.EBL_Finke10(),
+        photonField.EBL_Dominguez11(),
```
Comment (rafaelab) [inline calc_pairproduction.py:136]: Same as before: consider generating tables also for Domínguez upper/lower limits.

File: calc_photopionproduction.py [hunk 9]
```diff
@@ -8,21 +8,6 @@
 lgamma = np.linspace(6, 16, 251)  # tabulated Lorentz factors
 gamma = 10**lgamma
 
-fields = [
-    photonField.CMB(),
-    photonField.EBL_Kneiske04(),
-    photonField.EBL_Stecker05(),
-    photonField.EBL_Franceschini08(),
-    photonField.EBL_Finke10(),
-    photonField.EBL_Dominguez11(),
```
Comment (rafaelab) [inline calc_photopionproduction.py:17]: Same as before (compute tables for Domínguez upper/lower limit).

File: calc_photopionproduction.py [hunk 10]
```diff
@@ -36,8 +21,12 @@
 xs2 = d[1, :2049] * 1e-34  # [m^2]
 
 
-for field in fields:
-    print(field.name)
+def process(field):
+    """ 
+        calculate the interction rate on the given photonfield
```
Comment (rafaelab) [inline calc_photopionproduction.py:26]: ==> Calculate the interaction rate on the given photon field.

File: calc_photopionproduction.py [hunk 11]
```diff
@@ -91,3 +80,23 @@
...
     np.savetxt(fname, data, fmt=fmt, header=header)
+
+if __name__ == "__main__":
+    fields = [
+        photonField.CMB(),
+        photonField.EBL_Kneiske04(),
+        photonField.EBL_Stecker05(),
+        photonField.EBL_Franceschini08(),
+        photonField.EBL_Finke10(),
+        photonField.EBL_Dominguez11(),
```
Comment (rafaelab) [inline calc_photopionproduction.py:91]: Same as before (compute tables for Domínguez upper/lower limit).

File: calc_synchrotron.py [hunk 12]
```diff
@@ -38,8 +38,8 @@ def compute_spectrum(x, outputName):
       Wiley, 3rd ed., p. 681, eq. 14.91
 
     Input
-    . x: fraction between photon frequency and critical frequency
-    . outputName: name of output file
+      x:    fraction between photon frequency and critical frequency
```
Comment (rafaelab) [inline calc_synchrotron.py:41]: fraction between ==> ratio between

**Review by rafaelab — COMMENTED** [review]
Hi @JulienDoerner 
This PR looks good to me. There are a few typos that I pointed out. I also think it would be good to bring back the Domínguez11 lower/upper models, that were lost at some point.
(I see that @lukasmerten also left some comments.)

**Comment by JulienDoerner** [comment]
Hey @rafaelab, 
@lukasmerten has also worked on this topic and has a newer PR #17 opened. This work is included in Lukas PR and I will therefore close this one.

**Comment by rafaelab** [comment]
> Hey @rafaelab, @lukasmerten has also worked on this topic and has a newer PR #17 opened. This work is included in Lukas PR and I will therefore close this one.

I started revising from the oldest PR and just realised that, @JulienDoerner :p
